### ISSUE TITLE
Allow proxied @SpyBeans to be used with Mockito's inline mock maker

### ISSUE THREAD

**Issue by jacky1193610322** [spring-projects/spring-boot#22416]
spring-boot version is 2.1.3.RELEASE
I already addressed the issue in mockito. [issue](https://github.com/mockito/mockito/issues/1980)
the mockito 3.4.0 can mock static method, so I upgrade mockito to 3.4.0, but the mock framework changed, now when I use 
the @SpyBean and then verify it. it will throw NotAMockException.

```java
@SpringBootTest
@RunWith(SpringJUnit4ClassRunner.class)
@AutoConfigureMockMvc
public class AppGroupClusterActionTest {

  @Autowired
  private MockMvc mockMvc;

  @MockBean
  private AppGroupClusterService appGroupClusterService;

  @SpyBean
  private AppGroupClusterAction appGroupClusterAction;

  @Captor
  private ArgumentCaptor<Object[]> argumentCaptor;

  @Test
  public void testUpdateExecuteAuditSpel() throws Exception {
    ServiceResult mockRes = new ServiceResult();
    mockRes.setSuccess(true);
    doReturn(mockRes).when(appGroupClusterService).update(
        Mockito.any(), Mockito.any(), Mockito.any(),
        Mockito.any(), Mockito.any(), Mockito.any(),
        Mockito.any(), Mockito.any(), Mockito.any());

    mockMvc.perform(
        MockMvcRequestBuilders.post("/platforms/{platform}/appGroupNames/{appGroupName}/clusterNames/{clusterName}/update",
            "platform", "appGroupName", "clusterName").content("{}"))
        .andExpect(status().isOk())
        .andReturn();

//    not work
//    verify(appGroupClusterAction).addAfterAuditResult(argumentCaptor.capture());
    verify((AppGroupClusterAction) AopTestUtils.getTargetObject(appGroupClusterAction)).addAfterAuditResult(argumentCaptor.capture());
    assertArrayEquals(argumentCaptor.getValue(), new Object[]{"appGroupName", "platform", "clusterName", "{}"});

    verify((AppGroupClusterAction) AopTestUtils.getTargetObject(appGroupClusterAction)).addAfterAuditResult(argumentCaptor.capture());
    assertArrayEquals(argumentCaptor.getValue(), new Object[]{"appGroupName", "platform", "clusterName", "{}"});
  }
}
```

**Comment by wilkinsona** [comment]
@jacky1193610322 Thanks for the report. We don't officially support Mockito 3.4 yet, but it's likely that we will want to do so in Spring Boot 2.4. To help us to investigate your problem, can you please provide a minimal sample that reproduces it?

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by jacky1193610322** [comment]
sorry，I will offer simple code，but because of the rule of company, I can't upload the whole project, I will paste the core code.

**Comment by snicoll** [comment]
@jacky1193610322 it doesn't have to be your project. Actually, it has to be a minimal sample and that shouldn't expose any company specific information. Please don't paste the core code as the only thing we could do is try to copy/paste that in an actual project we can run.

**Comment by jacky1193610322** [comment]
I know. but push the code to GitHub is forbid.

**Comment by snicoll** [comment]
You can share the code in a different form (attaching a zip to this issue) or you can send it to me by email if that's easier. My email is on my github profile.

**Comment by jacky1193610322** [comment]
ok

**Comment by jacky1193610322** [comment]
sorry, the mail can't work now, can't upload code, but the example is simple, any aop proxy can reproduction.
Aspect
```java
@Component
@org.aspectj.lang.annotation.Aspect
public class Aspect {
  @Pointcut("execution(* com.example.demo.*.*(..))")
  public void pointcut() {
  }

  @Around("pointcut()")
  public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
    try {
      return joinPoint.proceed();
    } catch (Throwable e) {
      throw e;
    } finally {
      System.out.println("around");
    }
  }
}
```

http action
```java
@RequestMapping(value = "/demo/", produces = MediaType.APPLICATION_JSON_VALUE)
@RestController
public class Action {
  public void test() {

  }
}

test
```java
@SpringBootTest
class ActionTest {
  @SpyBean
  private Action action;

  @Test
  void name() {
    doNothing().when(action).test();
    Mockito.verify(action);
  }
}
```

**Comment by jacky1193610322** [comment]
sorry, the mail can't work now, can't upload code, but the example is simple, any aop proxy can reproduction.
Aspect
```java
@Component
@org.aspectj.lang.annotation.Aspect
public class Aspect {
  @Pointcut("execution(* com.example.demo.*.*(..))")
  public void pointcut() {
  }

  @Around("pointcut()")
  public Object around(ProceedingJoinPoint joinPoint) throws Throwable {
    try {
      return joinPoint.proceed();
    } catch (Throwable e) {
      throw e;
    } finally {
      System.out.println("around");
    }
  }
}
```

http action
```java
@RequestMapping(value = "/demo/", produces = MediaType.APPLICATION_JSON_VALUE)
@RestController
public class Action {
  public void test() {

  }
}

test
```java
@SpringBootTest
class ActionTest {
  @SpyBean
  private Action action;

  @Test
  void name() {
    doNothing().when(action).test();
    Mockito.verify(action);
  }
}
```

**Comment by wilkinsona** [comment]
@jacky1193610322 Your [example above](https://github.com/spring-projects/spring-boot/issues/22416#issuecomment-665555993) works for me with Mockito 3.4. I modified it slightly to correct the verification. This left the test looking like this:

```
@SpringBootTest
class ActionTest {
  @SpyBean
  private Action action;

  @Test
  void name() {
    doNothing().when(action).test();
    action.test();
    verify(action).test();
  }
}
```

It passes, outputting `around` to `System.out` numerous times as it does so.

Without a sample that reproduces the problem that you're seeing, I'm afraid we won't be able to make any progress here.

**Comment by jacky1193610322** [comment]
Thank you. 
```grovvy
plugins {
    id 'org.springframework.boot' version '2.3.2.RELEASE'
    id 'io.spring.dependency-management' version '1.0.9.RELEASE'
    id 'java'
}

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '1.8'

repositories {
    mavenCentral()
}

dependencies {
    implementation 'org.springframework.boot:spring-boot-starter-web'
    testImplementation('org.springframework.boot:spring-boot-starter-test') {
        exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
    }
    compile('org.aspectj:aspectjweaver:1.9.2')
    testCompile "org.mockito:mockito-core:3.4.4"
    testCompile "org.mockito:mockito-inline:3.4.4"
}

test {
    useJUnitPlatform()
}
```
you need add testCompile "org.mockito:mockito-inline:3.4.4"

**Comment by wilkinsona** [comment]
Thanks. I've reproduced the problem now. It isn't specific to Mockito 3.4.x. The same failure occurs with Mockito 3.3.3 with a dependency on `org.mockito:mockito-inline`.

**Comment by wilkinsona** [comment]
Looking more closely, I think this is a Mockito bug or at least a limitation. As a result of using `@SpyBean` a Mockito spy is created for `Action`. This spy is then proxied by Spring's AOP infrastructure and it's this proxy that is injected into the test class. It's then passed into a call to `when` at which point Mockito checks that it's a mock by checking that is has a non-null `MockHandler`. This works when using the standard mock maker but fails when using the inline mock maker. In the working case, `SubclassByteBuddyMockMaker` retrieves the `MockHandler` as follows:

```java
if (!(mock instanceof MockAccess)) {
    return null;
}
return ((MockAccess) mock).getMockitoInterceptor().getMockHandler();
```

`mock` is an instance of `MockAccess` and the handler from the interceptor is returned.

In the failing case, `InlineByteBuddyMockMaker` attempts to retrieve the `MockHandler` has follows:

```java
MockMethodInterceptor interceptor = mocks.get(mock);
if (interceptor == null) {
    return null;
} else {
    return interceptor.handler;
}
```

`mocks.get(mock)` returns `null` so `null` is returned. Given the same input to their `getHandler` methods, I would expect `InlineByteBuddyMockMaker` and `SubclassByteBuddyMockMaker ` to return the same output.

**Comment by raphw** [comment]
Cross-posting from our issue tracker: TLNR; this is not a Mockito bug, it's Spring relying on an undocumented implementation detail that might just change in any bugfix release.

The reason things currently (!) work with the `SubclassByteBuddyMockMaker` is that Spring proxies the (internal) `MockAccess` interface which we implement in the mock to extract the mock state which is stored in a field of the mock instance. We could just as much had read that field directly to avoid the interface; and this wouldn't even be such a bad idea as it less pollutes the reflection API. But at this time, when we call our internal interface method, it gets redirected to the actual mock object by Spring which is why this functions (accidentally).

The reason it does not work with the `InlineByteBuddyMockMaker` is that we cannot implement interfaces in a class after it has been loaded. The whole premise of the inline mock maker is to keep changes invisible to make use of the JVMs retransformation facilities even after class loading which makes this a requirement. Therefore, we need to identify mocks by the object's identity and store the mock state externally in a weak map (which requires using object identity), which is not the same for the Spring proxy object. We cannot change this and from our perspective nothing is broken.

From Mockito's side things are simple: if one does not provide "our" mock object to "our" API, we do not guarantee to function, not in the past, nor the future. That it does currently work with the subclass mock maker facility is rather accidental and might change in the future. If you implemented the `MockAccess` interface yourself and returned another object's mock state from it and passed this to Mockito, you'd break Mockito, too. If Spring cannot do anything about this, users need to unproxy the objects explicitly but there is nothing we can do about it. A proxied mock is not a mock from our side, just some object with a reference to a mock and of those there are of course plenty.

**Comment by wilkinsona** [comment]
It looks like we'll be able to address this with a custom `MockResolver` that builds on the change that @raphw has prototyped for https://github.com/mockito/mockito/issues/1980.

**Comment by raphw** [comment]
Great. new Mockito release to Central is underway just now.

**Comment by TimvdLippe** [comment]
I just realized we actually didn't publish a new minor version to Maven Central. However, I just tried to publish 3.6.0 and Bintray had issues. Hopefully we can figure out soon how to resolve that.

**Comment by TimvdLippe** [comment]
We have been able to fix our release woes and 3.6.0 has been published to Maven Central: https://repo1.maven.org/maven2/org/mockito/mockito-core/3.6.0/

**Comment by snicoll** [comment]
Thanks @TimvdLippe. The upgrade is scheduled for `2.4.0-RC1`, see #23924

**Comment by wilkinsona** [comment]
Closed by d9084ea.

**Comment by elxnis** [comment]
Hi @wilkinsona,

I think ```org.springframework.test.util.AopTestUtils#getUltimateTargetObject``` causes issues with scoped beans, similarly as before in https://github.com/spring-projects/spring-boot/issues/19020. I will try to test with my sample from that issue and raise new case if it is failing again.

Thanks,
Reinis
